### PR TITLE
w3c(phase 3): ensure tracecontext headers always take precedence

### PR DESCRIPTION
**Pull request by mabdinur**
## Description

Ensures when trace headers with the same trace_id but different span_ids are received by an application the tracecontext headers are treated as the source of truth.

The traceparent header will be used to set the parent_id
If the conflicting trace information was extracted from datadog headers, the span_id from the datadog headers will be stored in the _dd.parent_id parent_id tag.
All other propagation styles are third class citizens. If they conflict with tracecontext headers their span_ids will be overwritten.
This behavior will be validated by the following system-tests: https://github.com/DataDog/system-tests/pull/2385

### Reviewer checklist
- [ ] Test coverage seems ok.
- [ ] Appropriate labels assigned.

### CODE REVIEW

File: ext/distributed_tracing_headers.c [hunk 1]
```diff
@@ -396,6 +396,17 @@ ddtrace_distributed_tracing_result ddtrace_read_distributed_tracing_ids(ddtrace_
                     zend_hash_destroy(&result.tracestate_unknown_dd_keys);
                     result.tracestate_unknown_dd_keys = new_result.tracestate_unknown_dd_keys;
                     zend_hash_init(&new_result.tracestate_unknown_dd_keys, 0, NULL, NULL, 0);
+                    if (result.parent_id != new_result.parent_id) {
+                        // The span id from tracecontext takes precendence over all other headers
+                        result.parent_id = new_result.parent_id;
+                        // set last datadog span_id tag
+                        zval **lp_id;
+                        if (zend_hash_index_find(new_result.propagated_tags, ZEND_STRL("_dd.parent_id"), (void **) &lp_id) == SUCCESS) {
```
Comment (mabdinur) [inline ext/distributed_tracing_headers.c:404]:
```suggestion
                        if (zend_hash_index_find(new_result.propagated_tags, ZEND_STRL("_dd.parent_id"), (void **) &lp_id) == SUCCESS && lp_id != ZEND_STRL("0000000000000000")) {
```

File: ext/distributed_tracing_headers.c [hunk 2]
```diff
@@ -396,6 +396,17 @@ ddtrace_distributed_tracing_result ddtrace_read_distributed_tracing_ids(ddtrace_
...
                     result.tracestate_unknown_dd_keys = new_result.tracestate_unknown_dd_keys;
                     zend_hash_init(&new_result.tracestate_unknown_dd_keys, 0, NULL, NULL, 0);
+                    if (result.parent_id != new_result.parent_id) {
+                        // The span id from tracecontext takes precendence over all other headers
+                        result.parent_id = new_result.parent_id;
+                        // set last datadog span_id tag
+                        zval **lp_id;
+                        if (zend_hash_index_find(new_result.propagated_tags, ZEND_STRL("_dd.parent_id"), (void **) &lp_id) == SUCCESS && lp_id != ZEND_STRL("0000000000000000")) {
+                            zend_hash_str_update(result.propagated_tags, ZEND_STRL("_dd.parent_id"), &lp_id);
+                        else if (result_dd = ddtrace_read_distributed_tracing_ids_datadog(read_header, data) && result_dd.parent_id != 0) {
```
Comment (mabdinur) [inline ext/distributed_tracing_headers.c:406]:
```suggestion
                        } else if (result_dd = ddtrace_read_distributed_tracing_ids_datadog(read_header, data) && result_dd.parent_id != 0) {
```

File: tests/ext/distributed_tracing/distributed_trace_conflicting_span_ids.phpt [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+]);
+var_dump(DDTrace\generate_distributed_tracing_headers(["tracecontext"]));
+
+?>
+--EXPECT--
+array(2) {
+  ["traceparent"]=>
+  string(55) "00-0000000000000000000000000000002a-0000000000000001-01"
+  ["tracestate"]=>
+  string(24) "dd=p:000000000000000a;s:2;t.dm:-0"
```
Comment (mabdinur) [inline tests/ext/distributed_tracing/distributed_trace_conflicting_span_ids.phpt:22]:
```suggestion
  ["tracestate"]=>
  string(24) "dd=p:000000000000000a;s:2;t.dm:-0"
```

File: tests/ext/distributed_tracing/distributed_trace_conflicting_span_ids.phpt [hunk 4]
```diff
@@ -0,0 +1,21 @@
+--TEST--
+Test consume_distributed_tracing_headers() with array argument
+--ENV--
+DD_TRACE_GENERATE_ROOT_SPAN=0
```
Comment (mabdinur) [inline tests/ext/distributed_tracing/distributed_trace_conflicting_span_ids.phpt:4]:
```suggestion
DD_TRACE_GENERATE_ROOT_SPAN=0
DD_TRACE_PROPAGATION_STYLE=datadog,tracecontext
```

**Review by bwoebi — APPROVED** [review]
Looks good now :-)

**Pull request merged**